### PR TITLE
[Blazor] SupplyParameterFromForm dictionary support

### PR DESCRIPTION
**Pull request by javiercn**
https://github.com/dotnet/aspnetcore/issues/46688#issuecomment-1570315950

Performance data comparison with MVC. Without optimizations:

|                             Method | DictionarySize |        Mean |     Error |     StdDev |      Median |     Op/s | Ratio | RatioSD | Gen 0 | Gen 1 | Gen 2 | Allocated |
|----------------------------------- |--------------- |------------:|----------:|-----------:|------------:|---------:|------:|--------:|------:|------:|------:|----------:|
|        **ModelBinding_Dictionary_Mvc** |              **0** |    **37.24 μs** |  **1.225 μs** |   **3.515 μs** |    **36.80 μs** | **26,849.8** |  **1.00** |    **0.00** |     **-** |     **-** |     **-** |      **3 KB** |
| ModelBinding_Dictionary_Components |              0 |    25.46 μs |  2.157 μs |   5.904 μs |    23.65 μs | 39,273.2 |  0.69 |    0.16 |     - |     - |     - |      2 KB |
|                                    |                |             |           |            |             |          |       |         |       |       |       |           |
|        **ModelBinding_Dictionary_Mvc** |              **1** |    **76.61 μs** |  **3.605 μs** |  **10.110 μs** |    **74.20 μs** | **13,052.6** |  **1.00** |    **0.00** |     **-** |     **-** |     **-** |      **6 KB** |
| ModelBinding_Dictionary_Components |              1 |    32.89 μs |  1.215 μs |   3.485 μs |    31.90 μs | 30,400.0 |  0.43 |    0.08 |     - |     - |     - |      2 KB |
|                                    |                |             |           |            |             |          |       |         |       |       |       |           |
|        **ModelBinding_Dictionary_Mvc** |             **10** |    **94.69 μs** |  **1.904 μs** |   **5.116 μs** |    **94.05 μs** | **10,560.3** |  **1.00** |    **0.00** |     **-** |     **-** |     **-** |     **10 KB** |
| ModelBinding_Dictionary_Components |             10 |    40.79 μs |  1.653 μs |   4.717 μs |    39.50 μs | 24,515.6 |  0.43 |    0.05 |     - |     - |     - |      5 KB |
|                                    |                |             |           |            |             |          |       |         |       |       |       |           |
|        **ModelBinding_Dictionary_Mvc** |            **100** |   **362.88 μs** | **14.591 μs** |  **38.948 μs** |   **355.40 μs** |  **2,755.8** |  **1.00** |    **0.00** |     **-** |     **-** |     **-** |     **63 KB** |
| ModelBinding_Dictionary_Components |            100 |   120.84 μs |  4.973 μs |  13.945 μs |   119.90 μs |  8,275.4 |  0.33 |    0.04 |     - |     - |     - |     28 KB |
|                                    |                |             |           |            |             |          |       |         |       |       |       |           |
|        **ModelBinding_Dictionary_Mvc** |           **1000** | **2,952.31 μs** | **85.899 μs** | **227.791 μs** | **2,919.80 μs** |    **338.7** |  **1.00** |    **0.00** |     **-** |     **-** |     **-** |    **593 KB** |
| ModelBinding_Dictionary_Components |           1000 |   646.13 μs | 31.573 μs |  89.051 μs |   624.30 μs |  1,547.7 |  0.21 |    0.03 |     - |     - |     - |    175 KB |

### CODE REVIEW

File: src/Components/Endpoints/src/Binding/Factories/DictionaryConverterFactory.cs [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+        // Well-known dictionary types
+        // IDictionary<TKey, TValue>
+        // IReadOnlyDictionary<TKey, TValue>
+        // Dictionary<TKey, TValue>
+        // SortedDictionary<TKey, TValue>
+        // SortedList<TKey, TValue>
+        // ConcurrentDictionary<TKey, TValue>
+        // ImmutableDictionary<TKey, TValue>
+        // ImmutableSortedDictionary<TKey, TValue>
+
```
Comment (javiercn) [inline src/Components/Endpoints/src/Binding/Factories/DictionaryConverterFactory.cs:23]:
```suggestion
```

File: src/Components/Endpoints/test/Binding/FormDataMapperTests.cs [hunk 2]
```diff
@@ -487,7 +486,222 @@ public void CanDeserialize_Collections_CustomCollection()
...
+    }
+
+    [Fact]
+    public void CanDeserialize_Collections_CustomCollection()
+    {
+        // Arrange
+        var expected = new CustomCollection<int> { 10, 11, 12, 13, 14, 15, 16, 17, 18, 19 };
+        CanDeserialize_Collection<CustomCollection<int>, CustomCollection<int>, int>(expected);
+    }
+
```
Comment (javiercn) [inline src/Components/Endpoints/test/Binding/FormDataMapperTests.cs:704]:
```suggestion
```

File: src/Components/Endpoints/src/Binding/Converters/DictionaryConverter.cs [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+        bool foundCurrentValue;
+        bool currentElementSuccess;
+        bool succeded = true;
+
+        found = context.GetKeys().GetEnumerator().MoveNext();
+        if (!found)
+        {
+            result = default!;
+            return true;
+        }
```
Comment (MackinnonBuck) [inline src/Components/Endpoints/src/Binding/Converters/DictionaryConverter.cs:43]: Probably not a big deal, but I wonder if we could avoid the extra enumerator allocation by instead setting `found` in the loop below and creating the buffer lazily.

**Review by MackinnonBuck — APPROVED** [review]
Looks great to me 👍

File: src/Components/Endpoints/src/Binding/Converters/DictionaryConverter.cs (hunk 3 above)
Comment (javiercn) [reply src/Components/Endpoints/src/Binding/Converters/DictionaryConverter.cs:43]: I've removed the allocation in a separate change that removes all allocations in this code-path where I implement the struct enumerator pattern.

**Pull request merged**